### PR TITLE
Adapt dispatch region creation to always clone metadata-like/simple operations

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Operations like linalg.tensor_reshape and linalg.init_tensor need to
be cloned within the created dispatch region operation always. This
makes the logic of legalizing the flow.dispatch.workgroups op more
involved since it needs to account for
1) The values that are defined outside of the dispatch region needs to
   account for operations that will be cloned.
2) The cloned operations need to be inserted at the point where it can
   use the mapped values from within the dispatch op for its operations.

### CODE REVIEW

**Review by benvanik — APPROVED** [review]

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 1]
```diff
@@ -387,22 +415,59 @@ static void legalizeDispatchWorkgroupOperands(
...
-        repl, SmallPtrSet<Operation *, 8>(uses.begin(), uses.end()));
+    map.map(operand, repl);
   }
 
-  // Reinsert and replace old BB args.
-  for (BlockArgument ba : block.getArguments().take_front(numOldBBArgs)) {
-    block.addArgument(ba.getType());
-    ba.replaceAllUsesWith(block.getArguments().back());
+  // The only existing argument are for the outputs. Just need to add a new
+  // argument for the outpts and remap the value to use the new argument.
```
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:422]: s/outpts/outputs/

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 2]
```diff
@@ -326,48 +326,77 @@ bool isAlwaysClonedIntoDispatchOp(Operation *op) {
...
+    }
+  }
+
+  // Remove the temporarily cloned operations.
+  for (auto tempClonedOp : tempClonedOps) {
+    tempClonedOp->erase();
+  }
+}
+
+static Operation *getInsertionPoint(Block &block, ArrayRef<Value> usedValues) {
```
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:370]: Documentation for this?

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 3]
```diff
@@ -326,48 +326,77 @@ bool isAlwaysClonedIntoDispatchOp(Operation *op) {
...
+    llvm::SetVector<Operation *> &clonedOps) {
+  mlir::getUsedValuesDefinedAbove(region, valuesDefinedAbove);
+  OpBuilder::InsertionGuard g(b);
+  b.setInsertionPointToStart(&region.front());
+  llvm::SetVector<Operation *> tempClonedOps;
+  for (Value operand : valuesDefinedAbove) {
+    Operation *definingOp = operand.getDefiningOp();
+    if (!definingOp || !isAlwaysClonedIntoDispatchOp(definingOp)) continue;
+    clonedOps.insert(definingOp);
+    // Temporarily clone this operation. This is just so that when the values
```
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:343]: Ouch. IIUC, we are trying to only select the subset of values defined above from ops not gonna be copied into the region. What about using `mlir::visitUsedValuesDefinedAbove`, which has a callback to filter out values coming from `isAlwaysClonedIntoDispatchOp`?

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 4]
```diff
@@ -387,22 +415,59 @@ static void legalizeDispatchWorkgroupOperands(
...
-    ba.replaceAllUsesWith(block.getArguments().back());
+  // The only existing argument are for the outputs. Just need to add a new
+  // argument for the outpts and remap the value to use the new argument.
+  for (auto ba : block.getArguments().take_front(numOldBBArgs)) {
+    assert(ba.getType().isa<IREE::Flow::DispatchOutputType>());
+    map.map(ba, block.addArgument(ba.getType()));
+  }
+
+  auto getUsesOfValueOutsideOfDispatchOp =
+      [&](Value v) -> SmallPtrSet<Operation *, 4> {
```
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:429]: Nit: I think there is no need to explicitly list `-> SmallPtrSet<Operation *, 4>`? The logic of this lambda is fairly simple so missing that sounds okay to me.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 5]
```diff
@@ -387,22 +415,59 @@ static void legalizeDispatchWorkgroupOperands(
...
+  }
+
+  // Clone the marked operations.
+  for (Operation *op : clonedOps) {
+    SmallVector<Value, 2> clonedOpOperands = llvm::to_vector<2>(llvm::map_range(
+        op->getOperands(), [&map](Value v) { return map.lookup(v); }));
+    Operation *insertionPoint = getInsertionPoint(block, clonedOpOperands);
+    if (!insertionPoint) {
+      return op->emitOpError(
+          "failed to clone operation into dispatch workgroup op");
```
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:448]: Be a bit specific can be quite helpful when we hit this. What about something like "unexpected multiple blocks inside dispatch workgroups op"?

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:343]: Well, I moved away from this. Updated the approach to use a worklist and backtracking.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 5 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:448]: Thats not the reason for the failure. I actually dont think there should be a reason for failure here. Updated message to say failed to find insertion point.

**Pull request merged**